util/network-devp2p/src/ip_utils.rs

```diff
@@ -358,6 +365,96 @@ pub fn map_external_address(local: &NodeEndpoint) -> Option<NodeEndpoint> {
 	None
 }
 
+fn search_natpmp(local: &NodeEndpoint) -> Option<NodeEndpoint> {
+	if let SocketAddr::V4(ref local_addr) = local.address {
+		let local_port = local_addr.port();
+		let local_udp_port = local.udp_port;
+
+		let search_gateway_child = ::std::thread::spawn(move || {
+			let get_public_addr = |n: &mut Natpmp| {
+				// this function call want to receive `Response::Gateway` response from router, if other then it is an Error.
+				n.send_public_address_request()?;
+				::std::thread::sleep(Duration::from_millis(NAT_PMP_PORT_MAPPING_WAITING_DURATION));
+				match n.read_response_or_retry() {
+					Ok(Response::Gateway(gw)) => Ok(gw),
+					Err(e) => {
+						debug!("IP request error: {}", e);
+						Err(e)
+					},
+					_ => Err(natpmp::Error::NATPMP_ERR_UNDEFINEDERROR.into())
+				}
+			};
+
+			let get_mapped_tcp_port = |n: &mut Natpmp| {
+				// this function call want to receive `Response::TCP` response from router, if other then it is an Error.
+				n.send_port_mapping_request(Protocol::TCP, local_port, local_port, NAT_PMP_PORT_MAPPING_LIFETIME)?;
+				::std::thread::sleep(Duration::from_millis(NAT_PMP_PORT_MAPPING_WAITING_DURATION));
+				match n.read_response_or_retry() {
+					Ok(Response::TCP(tcp)) => Ok(tcp),
+					Err(e) => {
+						debug!("Port mapping for TCP error: {}", e);
+						Err(e)
+					},
+					_ => Err(natpmp::Error::NATPMP_ERR_UNDEFINEDERROR.into())
+				}
+			};
+
+			let get_mapped_udp_port = |n: &mut Natpmp| {
+				// this function call want to receive `Response::UDP` response from router, if other then it is an Error.
+				n.send_port_mapping_request(Protocol::UDP, local_udp_port, local_udp_port, NAT_PMP_PORT_MAPPING_LIFETIME)?;
+				::std::thread::sleep(Duration::from_millis(NAT_PMP_PORT_MAPPING_WAITING_DURATION));
+				match n.read_response_or_retry() {
+					Ok(Response::UDP(udp)) => Ok(udp),
+					Err(e) => {
+						debug!("Port mapping for UDP error: {}", e);
+						Err(e)
+					},
+					_ => Err(natpmp::Error::NATPMP_ERR_UNDEFINEDERROR.into())
+				}
+			};
+
+			match Natpmp::new() {
+				Ok(mut n) => {
+					let gw = get_public_addr(&mut n)?;
+					let tcp_r = get_mapped_tcp_port(&mut n)?;
+					let udp_r = get_mapped_udp_port(&mut n)?;
+
+					Ok(NodeEndpoint {
+						address: SocketAddr::V4(SocketAddrV4::new(*gw.public_address(), tcp_r.public_port())),
+						udp_port: udp_r.public_port()
+					})
+				},
+				Err(e) => Err(e)
+			}
```

Comment:
I'd suggest using `let mut n = Natpmp::new()?;` here, and remove the function closures (`get_mapped_udp_port` and such)

Reply:
done